Added support for building with MingW64 GCC on Windows

vityank opened the pull request:
This PR ads support for building with current latest MingW64 GCC and headers. Their headers lack some of SSPI and other stuff currently available in MS VS and Platform SDK's.

akallabeth commented on client/Windows/CMakeLists.txt:
```diff
@@ -18,6 +18,27 @@
 set(MODULE_NAME "wfreerdp-client")
 set(MODULE_PREFIX "FREERDP_CLIENT_WINDOWS_CONTROL")
 
+if (${CMAKE_GENERATOR} MATCHES "MinGW Makefiles")
```
this does not seem right, why is this necessary?

vityank replied:
Just to change the resource file. Cmake does not support 'if' inside set(), so forced to copy whole list with external if...

akallabeth commented on client/Windows/wfreerdp_mingw.rc:
```diff
@@ -0,0 +1,14 @@
+
```
This is exactly the same as `wfreerdp.rc`

vityank replied:
The content is same, but encoding is different. windres requires .rc files to be UTF-8, while MS suite uses .rc files encoded as UTF16+BOM. This is the only difference between those files.

akallabeth replied:
the `UTF16` might actually not be required

akallabeth commented on include/freerdp/channels/rdpdr.h:
```diff
@@ -298,6 +298,7 @@ enum FILE_FS_DEVICE_FLAG
 	FILE_DEVICE_SECURE_OPEN = 0x00000100
 };
 
+#if !defined(_WIN32) || !defined(__MINGW32__)
```
`_WIN32` should be defined by `MinGW` anyway, unnecessary change

vityank replied:
Just to check for Mingw only for Win32. Can reduce it actually as Mingw32 will probably not be defined by anything on Unix... It was more for safety.

akallabeth commented on winpr/include/winpr/intrin.h:
```diff
@@ -22,7 +22,7 @@
 #define WINPR_INTRIN_H
 
-#ifndef _WIN32
+#if !defined(_WIN32) || defined(__MINGW32__)
```
Same as above

akallabeth commented on winpr/include/winpr/nt.h:
```diff
@@ -1267,7 +1267,7 @@ static INLINE NTSTATUS NTSTATUS_FROM_WIN32(long x)
 #endif
 
-#ifdef _WIN32
+#if defined(_WIN32) && !defined(__MINGW32__)
```
Unnecessary

vityank replied:
This is needed - For MS tools you undefine and redeclare FILE_INFORMATION_CLASS and it's enum constants.
In MingW they already defined so all those tricks shall be disabled, with only winintr include remaining. Overwise you get double declaration errors for constants(You undefine only the enum itself, not constants, remember?)

akallabeth commented on winpr/include/winpr/nt.h:
```diff
@@ -1288,8 +1288,11 @@ static INLINE NTSTATUS NTSTATUS_FROM_WIN32(long x)
+#include <winternl.h>
 #endif
 
+#if !defined(_WIN32) || !defined(__MINGW32__)
```
Unnecessary

vityank replied:
Matches _FILE_INFORMATION_CLASS comment above.

akallabeth commented on winpr/include/winpr/nt.h:
```diff
@@ -1288,8 +1288,11 @@ static INLINE NTSTATUS NTSTATUS_FROM_WIN32(long x)
 #undef _FILE_INFORMATION_CLASS
 #undef FileDirectoryInformation
 
+#elif defined(_WIN32)
```
Will not work, `MinGW` does not define `WIN32` but it defines `_WIN32`

vityank replied:
It's _WIN32 already here.

akallabeth commented on winpr/include/winpr/pool.h:
```diff
@@ -110,7 +110,7 @@ typedef VOID (*PTP_WIN32_IO_CALLBACK)(PTP_CALLBACK_INSTANCE Instance, PVOID Cont
 #endif
 
-#if (!defined(_WIN32) || ((defined(_WIN32) && (_WIN32_WINNT < 0x0600))))
+#if (!defined(_WIN32) || ((defined(_WIN32) && (_WIN32_WINNT < 0x0600))) || ((defined(_WIN32) && defined(__MINGW32__))))
```
Unnecessary

vityank replied:
Mingw64 system headers/libraries doesn't have ThreadPool* API functions as of now. This is the only way to make it work(Using implementation of older APIs) until those functions will be added(Which I doubt of - they exist since Vista and still not there...).

akallabeth commented on winpr/libwinpr/sspi/NTLM/ntlm.h:
```diff
@@ -79,6 +79,7 @@ typedef enum _NTLM_STATE NTLM_STATE;
 
 enum _NTLM_AV_ID
 {
+#if !defined(_WIN32) || !defined(__MINGW32__)
```
Unnecessary

vityank replied:
Needed because constants below MsvAvTimestamp are defined in MingW64 headers as part of another system enum.
It's exists but not full, so we need this to add missing constants for MingW while not changing for MS Tools.

akallabeth commented on winpr/libwinpr/wtsapi/wtsapi.c:
```diff
@@ -52,9 +52,15 @@ static PWtsApiFunctionTable g_WtsApi = NULL;
 static HMODULE g_WtsApi32Module = NULL;
 static WtsApiFunctionTable WtsApi32_WtsApiFunctionTable = { 0 };
 
+#if defined(_WIN32) && defined(__MINGW32__)
```
Unnecessary

akallabeth commented on winpr/libwinpr/wtsapi/wtsapi.c:
```diff
@@ -52,9 +52,15 @@ static PWtsApiFunctionTable g_WtsApi = NULL;
 
+#if defined(_WIN32) && defined(__MINGW32__)
 #define WTSAPI32_LOAD_PROC(_name, _type)                                                   \
+	WtsApi32_WtsApiFunctionTable.p##_name = (_type)GetProcAddress(g_WtsApi32Module, "WT" \
```
This looks strange, this is `C` macro expansion, why is this necessary?

vityank replied:
MingW GCC 9 preprocessor dislikes the original declaration variant, and considers it a syntax error...

akallabeth replied:
Interesting, this should not happen, might be a undesired preprocessor behaviour, need to check this.

akallabeth commented on winpr/libwinpr/sspi/NTLM/ntlm.h:
```diff
@@ -87,6 +88,9 @@ enum _NTLM_AV_ID
 	MsvAvFlags,
 	MsvAvTimestamp,
+#elif defined(_WIN32)
+    MsvAvTimestamp = 7,
```
Formatting

vityank replied:
I'll take care of this then be back home.

vityank commented on channels/urbdrc/client/data_transfer.c:
```diff
@@ -1318,7 +1318,7 @@ static UINT urb_control_get_interface_request(IUDEVICE* pdev, URBDRC_CHANNEL_CAL
 	size_t out_size;
 	UINT32 InterfaceId, OutputBufferSize, usbd_status;
-	UINT16 interface;
+	UINT16 iface;
```
In GCC 9+ the 'interface' became keyword, as it really a keyword in newer C++ standards.
So compilation fails and it expects interface declaration in place to continue...

akallabeth replied:
Ok, then your compiler is set up wrong. this is `C` not `C++`

vityank replied:
Took quite a lot of time to find, what happens here.
MingW's core header has a definition in rpc.h(Included by ole2.h):

#ifndef __OBJC__
#undef interface
#define interface struct
#endif

and this causes the original bustage(Unrelated to GCC itself as I through initially).
C:\FreeRDP\channels\urbdrc\common\urbdrc_helpers.c:197:57: error: expected '{' before ',' token
  197 | const char* call_to_string(BOOL client, UINT32 interface, UINT32 functionId)
      |                                                         ^
In file included from C:/msys64/mingw64/x86_64-w64-mingw32/include/oleidl.h:7,
                 from C:/msys64/mingw64/x86_64-w64-mingw32/include/ole2.h:38,
                 from C:/msys64/mingw64/x86_64-w64-mingw32/include/wtypes.h:13,
                 from C:/FreeRDP/winpr/include/winpr/wtypes.h:55,
                 from C:\\FreeRDP\channels\urbdrc\common\urbdrc_helpers.h:24,
                 from C:\\FreeRDP\channels\urbdrc\common\urbdrc_helpers.c:20:

Renaming the variable as I did looks the best solution for this issue.

vityank commented:
On 11.05.2020 13:52, akallabeth wrote:
>
> *@akallabeth* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In channels/urbdrc/client/data_transfer.c 
> <https://github.com/FreeRDP/FreeRDP/pull/6158#discussion_r422954737>:
>
> > @@ -1318,7 +1318,7 @@ static UINT urb_control_get_interface_request(IUDEVICE* pdev, URBDRC_CHANNEL_CAL
>   {
>   	size_t out_size;
>   	UINT32 InterfaceId, OutputBufferSize, usbd_status;
> -	UINT16 interface;
> +	UINT16 iface;
>
> Ok, then your compiler is set up wrong. this is |C| not |C++|
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/FreeRDP/FreeRDP/pull/6158#discussion_r422954737>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AEMMG76FTTR2CX7VPMCSV23RQ7KGVANCNFSM4M45VOSA>.
>
Looks good:

-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: C:/msys64/mingw64/bin/gcc.exe
-- Check for working C compiler: C:/msys64/mingw64/bin/gcc.exe - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/msys64/mingw64/bin/g++.exe
-- Check for working CXX compiler: C:/msys64/mingw64/bin/g++.exe - works

Also process in task amanger is gcc.exe, not g++.exe, so looks good too.

Error by compiler with original 'interface':

FreeRDP\channels\urbdrc\common\urbdrc_helpers.c:200:29: error: expected 
expression before 'struct'
   200 |  const UINT32 interfaceId = interface & INTERFACE_ID_MASK;
       |                             ^~~~~~~~~

vityank commented:
On 11.05.2020 13:56, akallabeth wrote:
>
> *@akallabeth* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In client/Windows/wfreerdp_mingw.rc 
> <https://github.com/FreeRDP/FreeRDP/pull/6158#discussion_r422957031>:
>
> > @@ -0,0 +1,14 @@
> +
>
> the |UTF16| might actually not be required
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/FreeRDP/FreeRDP/pull/6158#discussion_r422957031>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AEMMG75NTMA3BOXMLRBCJILRQ7KWNANCNFSM4M45VOSA>.
>
Removed the wfreerdp_mingw.rc and if inside CMakeLists and changed the 
regular .rc file to ASCII encoding.

akallabeth commented:
@vityank currently a bit busy, so this will take a few days until I can test and comment again, sorry

vityank commented:
On 12.05.2020 13:46, akallabeth wrote:
>
> @vityank <https://github.com/vityank> currently a bit busy, so this 
> will take a few days until I can test and comment again, sorry
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/FreeRDP/FreeRDP/pull/6158#issuecomment-627264168>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AEMMG75XOK63QIOBI6YGIUDRRESG3ANCNFSM4M45VOSA>.
>
No problem. Thanks.

akallabeth commented:
@freerdp-bot test

vityank commented:
Updated the diffs.

Commit b138c0df2fec2ad88a3689f58dabf5c7a3954995 can be freely applied to 2.x tree and it's same as original diff, only without some SSPI changes, as MingW got their support extended lately.

Commit 5934508ee5fde4f13de811d205f4b77e295c3763 is solely for 3.x devel tree as it uses API's not available in pre 9.x CRT which is used by default on MingW-W64.

akallabeth commented:
@freerdp-bot test

Biswa96 commented:
@vityank Hi! Thank you for this approach. I think this change adds many burden in this project. I occasionally contribute to mingw-w64 and msys2. I think I can upstream the required changed in mingw-w64 headers. And msys2 has Ninja build tool. So, I think we can skip most of the changes here and freerdp devs don't have take mingw as special case.

vityank commented:
@Biswa96 Hi. Adding the missing parts to the compiler headers(Especially SSP stuff) and fixing the 'iface' keyword usage in there can be great. I already removed one change on latest PR updated because MingW got it fixed.
Because I don't track the MingW64 development directly, it would be nice if you'll post notification here then it's available in Msys2 repos.

Biswa96 commented:
I have a PR pending here. If it got merged I will create a package in msys2. Due to unavailability of actual remote machine, I have tested wfreerdp client with VirtualBox Host-Only and Bridged network interfaces. The only thing I need is some lab rats :)

All those SSP stuff are in mingw-w64 v9.0.0 which had been released tomorrow.

Biswa96 commented:
FreeRDP package is now available in msys2 repository now, link https://packages.msys2.org/base/mingw-w64-freerdp.

With some patches here https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-freerdp. Hopefully those can be upstreamed in proper way.

hardening commented:
AFAICT most of FreeRDP components build with mingw now, can we close this PR ?
